### PR TITLE
Megatron encoder-decoder refactor

### PR DESCRIPTION
**Pull request by michalivne**
This PR refactor Megatron T5 to a modular base encoder-decoder class hierarchy, and a T5 class which customizes the base class. The goal of the refactor is to support multiple encoder-decoder models with minimal code redundancy. 

# Refactor High Level Description

Hierarchy tree is expressed below via list alignment (i.e., right alignment for child class).

* `MegatronModule` sub-classes:
	* `MegatronTransformerEncoderModule`/`MegatronTransformerDecoderModule` - various architectures for manipulation of hidden --> hidden
	* `MegatronTransformerEncoderDecoderModule` - contains above classes and generates required joint encoder/decoder mask given independent masks, and reparametrization (e.g., vae)
	* `TokensEncoderDecoderModule` - modality-specific (e.g., tokens-tokens) model which map between pre-processed data (e.g., token ids) and encoder/decoder continuous input (i.e., hidden); handles embedders (i.e., embedding + positional embedding); handles projection from hidden to data distribution (e.g., project from hidden dimensions to the number of tokens)
* `NLPModel` sub-classes:
	* `MegatronBaseModel` - common initialization of Megatron
		* `MegatronLMEncoderDecoderModel` - modality-specific (e.g., text-text) model which pre-process raw data (e.g., tokenizer pre-process text to token ids); generic modality-specific training/testing/inference/loss methods
			* `MegatronT5Model` - a specific model (i.e., T5) which prepares datasets (i.e., with corresponding data augmentation), and model-specific features (e.g., sentinels in tokenizer)

# Notable Changes

1. Encoder/decoder mask is 2D with `1` for active element, and `0` for masked element (i.e., as opposed to 3D Megatron  mask with `False` for active and `True` for masked). This allows easy manipulation of masks. Conversion to 3D Megatron masks is done on-the-fly.
2. Joint cross-attention mask is not precomputed, but is rather done on-the-fly.

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+
+class MegatronLMEncoderDecoderModule(MegatronBaseModel):
+    """
+    Megatron encoder-decoder base class
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        super().__init__(cfg, trainer=trainer)
+
+        self.tokenizer = get_nmt_tokenizer(
```
Comment (mlgill) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:55]: @michalivne Would it be possible to move the tokenizer creation to the `MegatronT5Model` class? As currently implemented, `MegatronLMEncoderDecoderModule` is specific to NMT because of `get_nmt_tokenizer`), and it appears to be the only thing imposing this limitation -- the rest of the code is generally applicable. Because it is written in such a nice general style, I'd prefer to just subclass `MegatronLMEncoderDecoderModule`, but the tokenizer part will be a challenge. 

Seems like this could be true for other applications too, as I see that additional tokenizer getter functions exist.

Comment (mlgill) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:55]: The other option, to implement a new `__init__` function, will also not work for MegaMolBART because the model is dependent upon things initialized in `MegatronBaseModel` and `NLPModel`.

File: nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+            embedding_dropout_prob=hidden_dropout,
+        )
+        self.decoder_embedding = self.encoder_embedding
+        self._encoder_embedding_key = "encoder_embedding"
+        self._decoder_embedding_key = "decoder_embedding"
+
+        encoder = get_encoder_model(
+            arch=encoder_arch,
+            hidden_size=hidden_size,
+            ffn_hidden_size=hidden_size,
```
Comment (mlgill) [inline nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py:135]: @michalivne This is a bug -- should be `ffn_hidden_size=ffn_hidden_size,` as implemented, it prevents the feedforward layer size from expanding as predicted. 

Here is an example. - for these params:
```
 max_position_embeddings: 512
    hidden_dropout: 0.1
    kv_channels: null
    apply_query_key_layer_scaling: false
    persist_layer_norm: true
    layernorm_epsilon: 1.0e-05
    init_method_std: 0.02
    encoder_arch: transformer
    decoder_arch: transformer
    hidden_size: 36
    ffn_hidden_size: 144
    num_attention_heads: 3
    num_layers: 2
    vocab_size: 128
```

Here are the sizes of the layers:
```
[NeMo I 2022-02-05 20:36:47 test_model:48] enc_dec_model.encoder.model.layers.0.mlp.dense_h_to_4h.weight: torch.Size([36, 36])
[NeMo I 2022-02-05 20:36:47 test_model:48] enc_dec_model.encoder.model.layers.0.mlp.dense_h_to_4h.bias: torch.Size([36])
[NeMo I 2022-02-05 20:36:47 test_model:48] enc_dec_model.encoder.model.layers.0.mlp.dense_4h_to_h.weight: torch.Size([36, 36])
[NeMo I 2022-02-05 20:36:47 test_model:48] enc_dec_model.encoder.model.layers.0.mlp.dense_4h_to_h.bias: torch.Size([36])
```

Expected sizes are:
```
[NeMo I 2022-02-05 20:36:47 test_model:48] enc_dec_model.encoder.model.layers.0.mlp.dense_h_to_4h.weight: torch.Size([144, 36])
[NeMo I 2022-02-05 20:36:47 test_model:48] enc_dec_model.encoder.model.layers.0.mlp.dense_h_to_4h.bias: torch.Size([144])
[NeMo I 2022-02-05 20:36:47 test_model:48] enc_dec_model.encoder.model.layers.0.mlp.dense_4h_to_h.weight: torch.Size([36, 144])
[NeMo I 2022-02-05 20:36:47 test_model:48] enc_dec_model.encoder.model.layers.0.mlp.dense_4h_to_h.bias: torch.Size([36])
```

File: nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+            fp32_residual_connection=fp32_residual_connection,
+            activations_checkpoint_method=activations_checkpoint_method,
+            activations_checkpoint_num_layers=activations_checkpoint_num_layers,
+            layernorm_epsilon=layernorm_epsilon,
+            bias_gelu_fusion=bias_gelu_fusion,
+            persist_layer_norm=persist_layer_norm,
+            openai_gelu=openai_gelu,
+            onnx_safe=onnx_safe,
+            hidden_steps=hidden_steps,
+            hidden_blocks=hidden_steps,
```
Comment (mlgill) [inline nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py:158]: @michalivne This also appears to be a bug -- should be `hidden_blocks=hidden_blocks`.

File: nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+        )
+
+        decoder = get_decoder_model(
+            arch=decoder_arch,
+            hidden_size=hidden_size,
+            ffn_hidden_size=hidden_size,
```
Comment (mlgill) [inline nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py:164]: Same bug, should be `ffn_hidden_size=ffn_hidden_size`

File: nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+            fp32_residual_connection=fp32_residual_connection,
+            activations_checkpoint_method=activations_checkpoint_method,
+            activations_checkpoint_num_layers=activations_checkpoint_num_layers,
+            layernorm_epsilon=layernorm_epsilon,
+            bias_gelu_fusion=bias_gelu_fusion,
+            persist_layer_norm=persist_layer_norm,
+            openai_gelu=openai_gelu,
+            onnx_safe=onnx_safe,
+            hidden_steps=hidden_steps,
+            hidden_blocks=hidden_steps,
```
Comment (mlgill) [inline nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py:187]: Same bug, should be: `hidden_blocks=hidden_blocks`

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py (hunk 1 above)
Comment (michalivne) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:55]: Good idea. Will provide an easy override.

File: nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+
+    def forward(self, hidden_states, word_embeddings_weight):
+        output = parallel_lm_logits(hidden_states, word_embeddings_weight, self.parallel_output, bias=self.bias)
+        return output
+
+
+# TODO: add soft prompts
+
+
+class TokensEncoderDecoderModule(MegatronModule):
```
Comment (mlgill) [inline nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py:70]: Should this be in the modules folder?

File: nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py (hunk 4 above)
Comment (michalivne) [reply nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py:164]: Good catch. FIxed

File: nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py (hunk 5 above)
Comment (michalivne) [reply nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py:187]: Good catch.Fixed.

File: nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py (hunk 6 above)
Comment (michalivne) [reply nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py:70]: Worth discussion. Could either stay here (being specific to tokens, or moved to modules being rather generic.
@MaximumEntropy, @ericharper, @okuchaiev ?

File: nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py (hunk 2 above)
Comment (michalivne) [reply nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py]: Good catch. FIxed.

File: nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py (hunk 3 above)
Comment (michalivne) [reply nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py]: Good catch. Fixed.

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py (hunk 1 above)
Comment (michalivne) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py]: Done. Great feedback @mlgill! You can now override `_build_tokenizer` method in `MegatronLMEncoderDecoderModule`

Comment (mlgill) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py]: Looks great, thank you!

**Review by MaximumEntropy — DISMISSED** [review]
Super excited about all the stuff this PR opens up :) Just some minor discussion and suggestions on naming.

File: nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py (hunk 6 above)
Comment (michalivne) [reply nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py]: Done.

File: nemo/collections/nlp/data/language_modeling/megatron/t5_dataset.py [hunk 7]
```diff
@@ -289,7 +286,8 @@ def pad_and_convert_to_numpy(
     loss_mask = ([1] * num_tokens_dec) + ([0] * padding_length_dec)
     loss_mask = np.array(loss_mask, dtype=np.int64)
 
-    return tokens_enc, tokens_dec_in, labels, enc_mask, dec_mask, enc_dec_mask, loss_mask
+    # return tokens_enc, tokens_dec_in, labels, enc_mask, dec_mask, enc_dec_mask, loss_mask
```
[the remark this answers is not in the document]
Comment (michalivne) [reply nemo/collections/nlp/data/language_modeling/megatron/t5_dataset.py]: Done.

File: nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
```
[the remark this answers is not in the document]
Comment (michalivne) [reply nemo/collections/nlp/models/language_modeling/megatron/tokens_encoder_decoder.py]: I like it. Changed to `MegatronTokenLevelEncoderDecoderModule` to be consistent with other Megatron modules and models.

File: nemo/collections/nlp/modules/common/megatron/megatron_encoder_decoder.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    """Transformer encoder-decoder model.
+    """
+
+    def __init__(
+        self,
+        encoder,
+        decoder,
+        # AttnMaskType enum mask type (e.g., padding, casual)
+        encoder_attn_mask_type=None,
+        decoder_attn_mask_type=None,
```
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/megatron_encoder_decoder.py:32]: Is it possible to type hint this?

File: nemo/collections/nlp/modules/common/megatron/megatron_encoder_decoder.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+            except Exception as e:
+                raise ValueError("Failed inferring decoder_attn_mask_type, please provide AttnMaskType value")
+
+        self.encoder_attn_mask_type = encoder_attn_mask_type
+        self.decoder_attn_mask_type = decoder_attn_mask_type
+
+        self._encoder_key = "encoder"
+        self._decoder_key = "decoder"
+
+    # FIXME: no need to set decoder too?
```
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/megatron_encoder_decoder.py:56]: Still need this comment?

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

Had some minor comments, but will approve. Please change your copyrights to 2022.

File: nemo/collections/nlp/modules/common/megatron/megatron_encoder_decoder.py (hunk 10 above)
Comment (michalivne) [reply nemo/collections/nlp/modules/common/megatron/megatron_encoder_decoder.py:56]: Removed.

File: nemo/collections/nlp/modules/common/megatron/megatron_encoder_decoder.py (hunk 9 above)
Comment (michalivne) [reply nemo/collections/nlp/modules/common/megatron/megatron_encoder_decoder.py:32]: Done.

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!
